### DISCUSSION TITLE
Windows: don’t require admin/dev-mode for `bookmark set`

### DISCUSSION THREAD

**Discussion by maxiride** [jotaen/klog#378] [Something else]
I've added klog to my PATH and tried few commands. I can operate the logs, print them etc but the bookmark feature did not work. I'd love to provide further details but the error received is just that plain.

```
C:\_clienti>klog bookmark set timelog.klog
Error: Failed to create bookmark


C:\_clienti>klog version
Command line tool: v2.2  [0bb0d83]
File format: version 1 (RFC)

Checking for newer version...
You already have the latest version!
```

**Comment by jotaen** [comment]
Oh, that shouldn’t be… Which Windows version are you on? Could you check whether you have a hidden `.klog` folder in your home directory (`C:\Users\YOUR_USERNAME\.klog`).

For me it appears to work with Windows 10.

**Comment by jotaen** [comment]
Thinking out loud (a.k.a. note to self) it could be useful to have a `--debug` flag / `KLOG_DEBUG` env variable that provides more verbose error output.

**Comment by maxiride** [comment]
> Oh, that shouldn’t be… Which Windows version are you on? Could you check whether you have a hidden `.klog` folder in your home directory (`C:\Users\YOUR_USERNAME\.klog`).
> 
> For me it appears to work with Windows 10.

```
OS Name	Microsoft Windows 10 Pro
Version	10.0.19042 Build 19042
```
I do have the `.klog` folder in my home user directory, at the time of writing it is empty.

**Comment by jotaen** [comment]
I added support for the `KLOG_DEBUG` env variable to display more verbose error info. You can do:

- `KLOG_DEBUG=1 klog bookmark set foo.klg` with UNIX
- `set KLOG_DEBUG=1 && klog bookmark set foo.klg` with Windows

This is not released yet, though. In case you know how to build a Go project from sources you could try it already, that would help me. Otherwise you can wait for the next release, then we can get more info. (I still cannot reproduce this with Windows 10, unfortunately. The error originates [here](https://github.com/jotaen/klog/blob/main/src/app/context.go#L255).)

**Comment by maxiride** [comment]
> set KLOG_DEBUG=1 && klog bookmark set foo.klg

```
C:\Users\Federico\federicod\Projects\_clienti>set KLOG_DEBUG=1 && klog-debug bookmark set timelog.klog
Error: Failed to create bookmark


Original Error:
symlink C:\Users\Federico\federicod\Projects\_clienti\timelog.klog C:\Users\Federico/.klog/bookmark.klg: A required privilege is not held by the client.
```

Thanks to the infos in the error I circumvented the issue by running `cmd.exe` as Administrator and the process went through:

```
C:\Users\Federico\federicod\Projects\_clienti>set KLOG_DEBUG=1 && klog-debug bookmark set timelog.klog
Bookmarked file C:\Users\Federico\federicod\Projects\_clienti\timelog.klog
```

I honestly don't understand why an elevated shell is needed to create a symlink. I'm available to further guidance to provide more feedback on the issue.

**Comment by jotaen** [comment]
Terrific, thanks for checking it out with the debug flag! 🙏 

Is it possible that you had created a bookmark earlier with elevated permissions?

The issue seems to be this: [Symlink creation fails on Windows](https://github.com/golang/go/issues/22874#issuecomment-388455464) (with that privilege error) in case there is already a symlink at that location present. klog actually [removes the symlink before setting it](https://github.com/jotaen/klog/blob/main/src/app/context.go#L254), but it just swallows the error since I just assume that the error means “no such file”. This assumption is not correct, since there might be other causes of failure. This of course must cause subsequent failures, like the one that you reported.

(Just for context: bookmarking technically just means creating a symlink to the target file at `~/.klog/bookmark.klg`.)

Long story short: I made the error handling more robust: https://github.com/jotaen/klog/pull/68. Since you are apparently all set for compiling locally, could you do me a favour and try again with the `fix-symlink-windows` branch?

**Comment by maxiride** [comment]
> Is it possible that you had created a bookmark earlier with elevated permissions?

Impossible, I never used klog, and I was trying it for the first time and I never run a privileged shell if I don't intentionally have to.

---

1. I've completely deleted the `.klog` folder (as to make a clean environment for the app)
1. I've built the `fix-symlink-windows` and run it in an unprivileged shell, the symlink has not been created

```
C:\Users\Federico\federicod\Projects\_clienti>set KLOG_DEBUG=1 && klog-fix-symlink bookmark set timelog.klog
Error: Failed to create bookmark
Unable to create the new bookmark

Original Error:
symlink C:\Users\Federico\federicod\Projects\_clienti\timelog.klog C:\Users\Federico/.klog/bookmark.klg: A required privilege is not held by the client.

C:\Users\Federico\federicod\Projects\_clienti>
```

3. I've executed the above command in a privileged shell and the symlink was created

>  klog actually removes the symlink before setting it,

4. To check if the issue was in removing the symlink (on step 1 remember that I had deleted `.klog` so there was nothing to remove) I switched back to an unprivileged shell and tried to set again the bookmark without success, same error as above.

---

**TL;DR**

Seems to me that klog can successfully remove a previously created symlink (regardless of (un)privileged shell) but not create a new one.

**Comment by jotaen** [comment]
Ok, good to know that it’s not the remove step. Thanks for the detailed report!

I’m generally a bit shallow on Windows since I don’t use it regularly. I’ve tested the bookmark commands with both an admin account as well as a standard user one and couldn’t reproduce the issue. As far as Go is concerned there are no further options on [`os.Symlink`](https://golang.org/pkg/os/#Symlink), so it should just work out of the box (Go 1.19 and Win 10.19042.804).

Just to clarify: by “unprivileged shell” you mean you run the shell with a standard permission user account, or is there another way to configure permissions on Windows?

**Comment by jotaen** [comment]
Or does the target file have different permissions/ownership, and therefore it cannot symlink? (Just thinking out loud here…)

**Comment by maxiride** [comment]
> Just to clarify: by “unprivileged shell” you mean you run the shell with a standard permission user account, or is there another way to configure permissions on Windows?

I mean that that right click cmd.exe in start menu and click on "Run as administrator". My account (which is the only one) has the admin rights.

> Or does the target file have different permissions/ownership, and therefore it cannot symlink? (Just thinking out loud here…)

No clue on why it should have 🤔

**Comment by jotaen** [comment]
I can reproduce this now, the reason why it works for me is that my Windows VM has Developer mode turned on by default, which has more permissive settings. The root cause is some security mechanism in Windows that doesn’t allow the creation of symlinks for regular users. Generally it seems to me that symlink support on Windows by design is more [targeted at developers](https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/).

Not sure yet what the right move forward is, at least I couldn’t find a way to unblock an application easily (except running as admin, which is certainly not good).

**Comment by maxiride** [comment]
I'm glad you figured it out! It's a bummer this limitation.

> Not sure yet what the right move forward is, at least I couldn’t find a way to unblock an application easily (except running as admin, which is certainly not good).

Asking for elevated permissions isn't a good choice IMHO, a workaround, albeit anti-pattern, might be to create a configuration file in the `.klog` directory where to save the full path to the bookmarked journal.

Klog could then load the path and "under hood" append it to the -f arg, this way klog doesn't need to create a symlink while still always knowing where the journal is. The major con is that if the file the bookmark will be broken while on the other hand a symlonk will still work (I think, since it points to the inode block level)

Just a very quick brainstorming I thought to share.

**Comment by jotaen** [comment]
Yeah agreed, quite an annoyance… For the time being I [improved the error message on Windows](https://github.com/jotaen/klog/pull/73/files#diff-fdebd80862e1206e95a1164a7cb9cb9c00c89572ca8c020334d1ae0fee3b4060R23), so that it tells you that you need admin or dev-mode for the `bookmark set` command. This isn’t a solution of course, but it’s more user-friendly at least.

Solving this properly requires quite a bit of work unfortunately, not just to change the current implementation (e.g. to a text file that contains the path), but also to migrate existing bookmarks on the fly on Linux/Mac. I’d like to keep the platform-specific code as small as possible, so if I change the inner workings then I’d change it for all platforms.

Thanks for your assistance so far @maxiride

**Comment by jotaen** [comment]
This is fixed through https://github.com/jotaen/klog/commit/2c23b489ef0cbfa229c344375c5d9807a8f5e7f4 and will be shipped in the next release.

Internally, the bookmarks are stored as JSON file now (`~/.klog/bookmarks.json`). Existing symlink bookmarks will be migrated on-the-fly.
